haystack/utils/early_stopping.py

```diff
@@ -0,0 +1,88 @@
+from typing import Optional, Tuple, List, Dict
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class EarlyStopping:
+    """
+    Can be used to control early stopping with a Trainer class. A custom EarlyStopping class can be used instead as long
+    as it implements the method check_stopping and provides the attribute save_dir
+    """
+
+    def __init__(
+        self,
+        head: int = 0,
+        metric: str = "loss",
+        save_dir: Optional[str] = None,
+        mode: str = "min",
+        patience: int = 0,
+        min_delta: float = 0.001,
+        min_evals: int = 0,
+    ):
+        """
+        :param head: the prediction head referenced by the metric.
+        :param save_dir: the directory where to save the final best model, if None, no saving.
+        :param metric: The name of dev set metric to monitor (default: loss) which is extracted from the 0th prediction
```

Comment:
What are the alternatives to `loss` here? Let's list them out for users to choose from.

Reply:
It will depend on what metrics have been specified to calculate by the `Processor` (e.g. the `SquadProcessor`). So I can provide examples of what would be available by default for a Reader training, but they are not guaranteed to be there.

Reply:
Yeah I'm not entirely sure what to put here since it depends on the metrics that are defined by `head.metric` used in this function https://github.com/deepset-ai/haystack/blob/b03de53716b5b755a4fe593b3d0329a9ffdaea3a/haystack/modeling/evaluation/eval.py#L136

And all possible metrics in Haystack are defined in this file https://github.com/deepset-ai/haystack/blob/main/haystack/modeling/evaluation/metrics.py which I'm not sure has a docs page right now.

Reply:
I added some specific suggestions for `FARMReader` training and `DensePassageRetriever` training since those are the two classes that support using this `EarlyStopping` feature.